File: main/src/main/java/com/sedmelluq/discord/lavaplayer/tools/http/AbstractHttpContextFilter.java
package com.sedmelluq.discord.lavaplayer.tools.http;

import org.apache.http.HttpResponse;
import org.apache.http.client.methods.HttpUriRequest;
import org.apache.http.client.protocol.HttpClientContext;

public abstract class AbstractHttpContextFilter implements HttpContextFilter {
    private final HttpContextFilter delegate;

    protected AbstractHttpContextFilter(HttpContextFilter delegate) {
        this.delegate = delegate;
    }

    @Override
    public void onContextOpen(HttpClientContext context) {
        if (delegate != null) {
            delegate.onContextOpen(context);
        }
    }

    @Override
    public void onContextClose(HttpClientContext context) {
        if (delegate != null) {
            delegate.onContextClose(context);
        }
    }

    @Override
    public void onRequest(HttpClientContext context, HttpUriRequest request, boolean isRepetition) {
        if (delegate != null) {
            delegate.onRequest(context, request, isRepetition);
        }
    }

    @Override
    public boolean onRequestResponse(HttpClientContext context, HttpUriRequest request, HttpResponse response) {
        if (delegate != null) {
            return delegate.onRequestResponse(context, request, response);
        }

        return false;
    }

    @Override
    public boolean onRequestException(HttpClientContext context, HttpUriRequest request, Throwable error) {
        if (delegate != null) {
            return delegate.onRequestException(context, request, error);
        }

        return false;
    }
}

File: main/src/main/java/lavaplayer/tools/http/AbstractHttpContextFilter.kt
package lavaplayer.tools.http

import org.apache.http.HttpResponse
import org.apache.http.client.methods.HttpUriRequest
import org.apache.http.client.protocol.HttpClientContext

abstract class AbstractHttpContextFilter(private val delegate: HttpContextFilter?) : HttpContextFilter {
    override fun onContextOpen(context: HttpClientContext) {
        delegate?.onContextOpen(context)
    }

    override fun onContextClose(context: HttpClientContext) {
        delegate?.onContextClose(context)
    }

    override fun onRequest(context: HttpClientContext, request: HttpUriRequest, isRepetition: Boolean) {
        delegate?.onRequest(context, request, isRepetition)
    }

    override fun onRequestResponse(
        context: HttpClientContext,
        request: HttpUriRequest,
        response: HttpResponse
    ): Boolean {
        return delegate?.onRequestResponse(context, request, response) ?: false
    }

    override fun onRequestException(context: HttpClientContext, request: HttpUriRequest, error: Throwable): Boolean {
        return delegate?.onRequestException(context, request, error) ?: false
    }
}
